File: scripts/lib/dmi-bulk-storage.mjs
import fs from 'node:fs/promises';
import path from 'node:path';

export const DMI_BULK_STORAGE_SCHEMA = 'dmi-bulk-source-dictionary-v1';
export const DMI_BULK_MAX_STORED_BYTES = 256 * 1024 * 1024;
export const DMI_BULK_MAX_LEGACY_BYTES = 1536 * 1024 * 1024;
const MAX_SOURCE_RECORDS = 2_000_000;
const MAX_CONTAINER_NODES = 50_000_000;
const MAX_DEPTH = 64;
const GROUPS = Object.freeze(['asset', 'spatial', 'semantics']);
const REFERENCE_KEY = '$dmiSource';
const ASSET_FIELDS = new Set([
  'collection', 'modelRun', 'itemId', 'assetIdentitySha256', 'assetSizeBytes',
  'acquiredAt', 'contentLengthBytes', 'contentSha256', 'itemCreatedAt',
  'itemUpdatedAt', 'nativeValidTime',
]);
const SPATIAL_FIELDS = new Set([
  'gridPoint', 'gridDefinitionSha256', 'distanceKm', 'samplingPoint',
  'samplingPointSha256', 'samplingIdentitySha256', 'sourceRegistrySha256',
  'coastalPartId', 'partId', 'zoneId', 'entityId', 'entityType', 'parentZoneId',
  'samplingContext', 'physicalPointSha256', 'verticalLayer', 'verticalLayerRankM',
]);

function isObject(value) {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

function exactKeys(value, keys) {
  return isObject(value)
    && JSON.stringify(Object.keys(value).sort()) === JSON.stringify([...keys].sort());
}

function sourceContainers(document) {
  if (!isObject(document) || !isObject(document.zones)) {
    throw new Error('DMI bulk document has no zones object');
  }
  const result = [];
  for (const zone of Object.values(document.zones)) {
    if (!isObject(zone)) continue;
    const rows = isObject(zone.hourly)
      ? Object.values(zone.hourly) : Array.isArray(zone.hourly) ? zone.hourly : [];
    for (const row of rows) {
      if (isObject(row) && isObject(row.sources)) result.push(row.sources);
    }
  }
  return result;
}

function partition(source) {
  if (Object.hasOwn(source, REFERENCE_KEY)) {
    throw new Error('DMI source collides with reserved storage reference');
  }
  const entries = Object.fromEntries(GROUPS.map(group => [group, []]));
  for (const entry of Object.entries(source)) {
    const group = ASSET_FIELDS.has(entry[0]) ? 'asset' : SPATIAL_FIELDS.has(entry[0]) ? 'spatial' : 'semantics';
    entries[group].push(entry);
  }
  return Object.fromEntries(GROUPS.map(group => [group, Object.fromEntries(entries[group])]));
}

function deepFreeze(value, seen = new Set()) {
  if (!value || typeof value !== 'object' || seen.has(value)) return value;
  seen.add(value);
  for (const child of Object.values(value)) deepFreeze(child, seen);
  return Object.freeze(value);
}

function boundedTree(root) {
  const stack = [[root, 0]];
  let nodes = 0;
  let references = 0;
  while (stack.length) {
    const [value, depth] = stack.pop();
    nodes += 1;
    if (nodes > MAX_CONTAINER_NODES || depth > MAX_DEPTH) {
      throw new Error('DMI storage structure bound exceeded');
    }
    if (isObject(value)) {
      if (Object.hasOwn(value, REFERENCE_KEY)) references += 1;
      for (const child of Object.values(value)) stack.push([child, depth + 1]);
    } else if (Array.isArray(value)) {
      for (const child of value) stack.push([child, depth + 1]);
    }
  }
  return references;
}

function validateTables(tables) {
  if (!exactKeys(tables, GROUPS)) throw new Error('DMI source table shape is invalid');
  for (const group of GROUPS) {
    if (!Array.isArray(tables[group]) || tables[group].length > MAX_SOURCE_RECORDS) {
      throw new Error('DMI source table bound is invalid');
    }
    for (const row of tables[group]) {
      if (!isObject(row) || Object.hasOwn(row, REFERENCE_KEY)) {
        throw new Error('DMI source table row is invalid');
      }
      const keys = Object.keys(row);
      if ((group === 'asset' && keys.some(key => !ASSET_FIELDS.has(key)))
        || (group === 'spatial' && keys.some(key => !SPATIAL_FIELDS.has(key)))
        || (group === 'semantics' && keys.some(key => ASSET_FIELDS.has(key) || SPATIAL_FIELDS.has(key)))) {
        throw new Error('DMI source table partition is invalid');
      }
    }
  }
  return tables;
}

function jsonBytes(value) {
  return Buffer.byteLength(JSON.stringify(value));
}

export function decodeDmiBulkWrapper(wrapper, { storedBytes = 0, expandSources = true } = {}) {
  if (!exactKeys(wrapper, ['storageSchema', 'sourceTables', 'document'])
    || wrapper.storageSchema !== DMI_BULK_STORAGE_SCHEMA) {
    throw new Error('DMI storage wrapper identity is invalid');
  }
  const referencesInTree = boundedTree(wrapper.document);
  if (boundedTree(wrapper.sourceTables) !== 0) {
    throw new Error('DMI source table contains a storage reference');
  }
  const tables = validateTables(wrapper.sourceTables);
  const tableBytes = Object.fromEntries(GROUPS.map(group => [
    group, tables[group].map(jsonBytes),
  ]));
  let count = 0;
  let logicalUpperBound = Math.max(0, storedBytes);
  for (const sources of sourceContainers(wrapper.document)) {
    for (const component of Object.keys(sources)) {
      const reference = sources[component];
      if (!isObject(reference) || !exactKeys(reference, [REFERENCE_KEY])) {
        if (isObject(reference)) throw new Error('DMI encoded source reference is invalid');
        continue;
      }
      const indices = reference[REFERENCE_KEY];
      if (!Array.isArray(indices) || indices.length !== GROUPS.length) {
        throw new Error('DMI encoded source tuple is invalid');
      }
      const keys = new Set();
      let bodyBytes = 0;
      let nonemptyGroups = 0;
      for (let offset = 0; offset < GROUPS.length; offset += 1) {
        const group = GROUPS[offset];
        const index = indices[offset];
        if (!Number.isSafeInteger(index) || index < 0 || index >= tables[group].length) {
          throw new Error('DMI encoded source index is invalid');
        }
        const row = tables[group][index];
        for (const key of Object.keys(row)) {
          if (keys.has(key)) throw new Error('DMI encoded source fields overlap');
          keys.add(key);
        }
        const bytes = tableBytes[group][index];
        if (bytes > 2) { bodyBytes += bytes - 2; nonemptyGroups += 1; }
      }
      const mergedBytes = 2 + bodyBytes + Math.max(0, nonemptyGroups - 1);
      logicalUpperBound += Math.max(0, mergedBytes - jsonBytes(reference));
      if (logicalUpperBound > DMI_BULK_MAX_LEGACY_BYTES) {
        throw new Error('DMI decoded document exceeds its logical bound');
      }
      count += 1;
      if (count > MAX_SOURCE_RECORDS) throw new Error('DMI source record bound exceeded');
    }
  }
  if (count !== referencesInTree) {
    throw new Error('DMI source reference appeared outside an hourly source container');
  }
  if (expandSources) {
    for (const group of GROUPS) {
      for (const row of tables[group]) deepFreeze(row);
      Object.freeze(tables[group]);
    }
    Object.freeze(tables);
    for (const sources of sourceContainers(wrapper.document)) {
      for (const component of Object.keys(sources)) {
        const reference = sources[component];
        if (!isObject(reference) || !exactKeys(reference, [REFERENCE_KEY])) continue;
        const entries = GROUPS.flatMap((group, offset) => (
          Object.entries(tables[group][reference[REFERENCE_KEY][offset]])
        ));
        sources[component] = Object.freeze(Object.fromEntries(entries));
      }
    }
  }
  return wrapper.document;
}

export async function readDmiBulkDocument(file, { optional = false, expandSources = true } = {}) {
  const info = await fs.lstat(file).catch(error => {
    if (optional && error?.code === 'ENOENT') return null;
    throw error;
  });
  if (info === null) return null;
  if (!info.isFile() || info.isSymbolicLink() || info.size < 2 || info.size > DMI_BULK_MAX_LEGACY_BYTES) {
    throw new Error('DMI bulk input is not a bounded regular file');
  }
  if (info.size > 512 * 1024 * 1024) {
    throw new Error('Legacy DMI bulk input requires Python producer normalization before Node use');
  }
  let parsed;
  try { parsed = JSON.parse(await fs.readFile(file, 'utf8')); }
  catch { throw new Error('DMI bulk input cannot be parsed'); }
  if (!isObject(parsed)) throw new Error('DMI bulk input must be an object');
  if (Object.hasOwn(parsed, 'storageSchema')) {
    if (info.size > DMI_BULK_MAX_STORED_BYTES) {
      throw new Error('DMI encoded input exceeds its stored byte bound');
    }
    return decodeDmiBulkWrapper(parsed, { storedBytes: info.size, expandSources });
  }
  if (!isObject(parsed.zones)) throw new Error('DMI legacy bulk input has no zones object');
  return parsed;
}

function encodedView(document) {
  if (!isObject(document) || !isObject(document.zones)) {
    throw new Error('DMI bulk document has no zones object');
  }
  const tables = Object.fromEntries(GROUPS.map(group => [group, []]));
  const indexes = Object.fromEntries(GROUPS.map(group => [group, new Map()]));
  let count = 0;
  const zones = Object.fromEntries(Object.entries(document.zones).map(([zoneId, zone]) => {
    if (!isObject(zone)) return [zoneId, zone];
    const encodeRow = row => {
      if (!isObject(row) || !isObject(row.sources)) return row;
      const sources = Object.fromEntries(Object.entries(row.sources).map(([component, source]) => {
        if (!isObject(source)) return [component, source];
        const grouped = partition(source);
        const references = GROUPS.map(group => {
          const key = JSON.stringify(grouped[group]);
          let index = indexes[group].get(key);
          if (index === undefined) {
            index = tables[group].length;
            indexes[group].set(key, index);
            tables[group].push(grouped[group]);
          }
          return index;
        });
        count += 1;
        if (count > MAX_SOURCE_RECORDS) throw new Error('DMI source record bound exceeded');
        return [component, { [REFERENCE_KEY]: references }];
      }));
      return { ...row, sources };
    };
    const hourly = isObject(zone.hourly)
      ? Object.fromEntries(Object.entries(zone.hourly).map(([time, row]) => [time, encodeRow(row)]))
      : Array.isArray(zone.hourly) ? zone.hourly.map(encodeRow) : zone.hourly;
    return [zoneId, { ...zone, hourly }];
  }));
  const encodedDocument = { ...document, zones };
  return { storageSchema: DMI_BULK_STORAGE_SCHEMA, sourceTables: tables, document: encodedDocument };
}

export async function writeDmiBulkDocument(file, document) {
  const destination = path.resolve(file);
  const temporary = `${destination}.tmp`;
  const wrapper = encodedView(document);
  try {
    const bytes = Buffer.from(`${JSON.stringify(wrapper)}\n`, 'utf8');
    if (bytes.length > DMI_BULK_MAX_STORED_BYTES) {
      throw new Error('DMI encoded output exceeds its stored byte bound');
    }
    await fs.mkdir(path.dirname(destination), { recursive: true });
    await fs.writeFile(temporary, bytes);
    await fs.rename(temporary, destination);
    return bytes.length;
  } finally {
    await fs.rm(temporary, { force: true }).catch(() => {});
  }
}

File: scripts/lib/dmi_bulk_storage.py
"""Bounded, lossless storage codec for the private DMI bulk document."""
from __future__ import annotations

import contextlib
import copy
import json
import os
import pathlib
import shutil
import stat
from typing import Any, Iterator


STORAGE_SCHEMA = "dmi-bulk-source-dictionary-v1"
GROUPS = ("asset", "spatial", "semantics")
REFERENCE_KEY = "$dmiSource"
MAX_STORED_BYTES = 256 * 1024 * 1024
MAX_LEGACY_BYTES = 1536 * 1024 * 1024
MAX_SOURCE_RECORDS = 2_000_000
MAX_CONTAINER_NODES = 50_000_000
MAX_DEPTH = 64

ASSET_FIELDS = frozenset({
    "collection", "modelRun", "itemId", "assetIdentitySha256", "assetSizeBytes",
    "acquiredAt", "contentLengthBytes", "contentSha256", "itemCreatedAt",
    "itemUpdatedAt", "nativeValidTime",
})
SPATIAL_FIELDS = frozenset({
    "gridPoint", "gridDefinitionSha256", "distanceKm", "samplingPoint",
    "samplingPointSha256", "samplingIdentitySha256", "sourceRegistrySha256",
    "coastalPartId", "partId", "zoneId", "entityId", "entityType",
    "parentZoneId", "samplingContext", "physicalPointSha256", "verticalLayer",
    "verticalLayerRankM",
})


def _reject_non_finite(value: str) -> None:
    raise ValueError(f"non-finite JSON token rejected: {value}")


def _canonical_bytes(value: Any) -> bytes:
    return json.dumps(
        value, ensure_ascii=False, sort_keys=True, separators=(",", ":"),
        allow_nan=False,
    ).encode("utf-8")


def _source_containers(document: Any) -> Iterator[dict[str, Any]]:
    if not isinstance(document, dict) or not isinstance(document.get("zones"), dict):
        raise ValueError("DMI bulk document has no zones object")
    for zone in document["zones"].values():
        if not isinstance(zone, dict):
            continue
        hourly = zone.get("hourly")
        rows = (
            hourly.values() if isinstance(hourly, dict)
            else hourly if isinstance(hourly, list)
            else ()
        )
        for row in rows:
            if isinstance(row, dict) and isinstance(row.get("sources"), dict):
                yield row["sources"]


def _partition(source: dict[str, Any]) -> dict[str, dict[str, Any]]:
    if REFERENCE_KEY in source:
        raise ValueError("DMI source collides with reserved storage reference")
    return {
        "asset": {key: value for key, value in source.items() if key in ASSET_FIELDS},
        "spatial": {key: value for key, value in source.items() if key in SPATIAL_FIELDS},
        "semantics": {
            key: value for key, value in source.items()
            if key not in ASSET_FIELDS and key not in SPATIAL_FIELDS
        },
    }


@contextlib.contextmanager
def _encoded_view(document: dict[str, Any]):
    """Temporarily replace source leaves, restoring the caller object in all cases."""
    if ASSET_FIELDS & SPATIAL_FIELDS:
        raise RuntimeError("DMI source storage partitions overlap")
    tables: dict[str, list[dict[str, Any]]] = {group: [] for group in GROUPS}
    indexes: dict[str, dict[bytes, int]] = {group: {} for group in GROUPS}
    replacements: list[tuple[dict[str, Any], str, dict[str, Any]]] = []
    count = 0
    try:
        for sources in _source_containers(document):
            for component in tuple(sources):
                source = sources[component]
                if not isinstance(source, dict):
                    continue
                partitions = _partition(source)
                references: list[int] = []
                for group in GROUPS:
                    value = partitions[group]
                    key = _canonical_bytes(value)
                    index = indexes[group].get(key)
                    if index is None:
                        index = len(tables[group])
                        indexes[group][key] = index
                        tables[group].append(value)
                    references.append(index)
                replacements.append((sources, component, source))
                sources[component] = {REFERENCE_KEY: references}
                count += 1
                if count > MAX_SOURCE_RECORDS:
                    raise ValueError("DMI source record bound exceeded")
        yield {
            "storageSchema": STORAGE_SCHEMA,
            "sourceTables": tables,
            "document": document,
        }, count
    finally:
        for sources, component, source in reversed(replacements):
            sources[component] = source


def encode_dmi_bulk_document(document: dict[str, Any]) -> dict[str, Any]:
    """Return an independent encoded wrapper; intended for tests and tooling."""
    logical = copy.deepcopy(document)
    with _encoded_view(logical) as (wrapper, _count):
        return copy.deepcopy(wrapper)


def _bounded_tree(document: Any) -> int:
    stack = [(document, 0)]
    nodes = 0
    references = 0
    while stack:
        value, depth = stack.pop()
        nodes += 1
        if nodes > MAX_CONTAINER_NODES or depth > MAX_DEPTH:
            raise ValueError("DMI storage structure bound exceeded")
        if isinstance(value, dict):
            if REFERENCE_KEY in value:
                references += 1
            stack.extend((child, depth + 1) for child in value.values())
        elif isinstance(value, list):
            stack.extend((child, depth + 1) for child in value)
    return references


def _validate_tables(tables: Any) -> dict[str, list[dict[str, Any]]]:
    if not isinstance(tables, dict) or set(tables) != set(GROUPS):
        raise ValueError("DMI source table shape is invalid")
    for group in GROUPS:
        rows = tables[group]
        if not isinstance(rows, list) or len(rows) > MAX_SOURCE_RECORDS:
            raise ValueError("DMI source table bound is invalid")
        for value in rows:
            if not isinstance(value, dict) or REFERENCE_KEY in value:
                raise ValueError("DMI source table row is invalid")
            keys = set(value)
            if (
                (group == "asset" and not keys <= ASSET_FIELDS)
                or (group == "spatial" and not keys <= SPATIAL_FIELDS)
                or (group == "semantics" and bool(keys & (ASSET_FIELDS | SPATIAL_FIELDS)))
            ):
                raise ValueError("DMI source table partition is invalid")
    return tables


def decode_dmi_bulk_wrapper(
    wrapper: dict[str, Any], *, stored_bytes: int = 0, expand_sources: bool = True,
) -> dict[str, Any]:
    if not isinstance(wrapper, dict) or set(wrapper) != {
        "storageSchema", "sourceTables", "document",
    } or wrapper.get("storageSchema") != STORAGE_SCHEMA:
        raise ValueError("DMI storage wrapper identity is invalid")
    references_in_tree = _bounded_tree(wrapper["document"])
    if _bounded_tree(wrapper["sourceTables"]) != 0:
        raise ValueError("DMI source table contains a storage reference")
    tables = _validate_tables(wrapper["sourceTables"])
    document = wrapper["document"]
    table_bytes = {
        group: [len(_canonical_bytes(value)) for value in tables[group]]
        for group in GROUPS
    }
    source_count = 0
    logical_upper_bound = max(0, stored_bytes)
    # Pass one validates every reference and the complete expansion bound.
    # Nothing in the caller-owned wrapper is changed until this pass succeeds.
    for sources in _source_containers(document):
        for component in tuple(sources):
            reference = sources[component]
            if not isinstance(reference, dict) or set(reference) != {REFERENCE_KEY}:
                if isinstance(reference, dict):
                    raise ValueError("DMI encoded source reference is invalid")
                continue
            indices = reference[REFERENCE_KEY]
            if not isinstance(indices, list) or len(indices) != len(GROUPS):
                raise ValueError("DMI encoded source tuple is invalid")
            decoded_keys: set[str] = set()
            body_bytes = 0
            nonempty_groups = 0
            for group, index in zip(GROUPS, indices):
                if isinstance(index, bool) or not isinstance(index, int) \
                        or not 0 <= index < len(tables[group]):
                    raise ValueError("DMI encoded source index is invalid")
                value = tables[group][index]
                if decoded_keys & value.keys():
                    raise ValueError("DMI encoded source fields overlap")
                decoded_keys.update(value)
                encoded_bytes = table_bytes[group][index]
                if encoded_bytes > 2:
                    body_bytes += encoded_bytes - 2
                    nonempty_groups += 1
            merged_bytes = 2 + body_bytes + max(0, nonempty_groups - 1)
            logical_upper_bound += max(0, merged_bytes - len(_canonical_bytes(reference)))
            if logical_upper_bound > MAX_LEGACY_BYTES:
                raise ValueError("DMI decoded document exceeds its logical bound")
            source_count += 1
            if source_count > MAX_SOURCE_RECORDS:
                raise ValueError("DMI decoded source record bound exceeded")
    if source_count != references_in_tree:
        raise ValueError("DMI source reference appeared outside an hourly source container")
    if not isinstance(document, dict):
        raise ValueError("DMI decoded document is invalid")
    if expand_sources:
        # Pass two materializes only after the complete wrapper is accepted.
        for sources in _source_containers(document):
            for component in tuple(sources):
                reference = sources[component]
                if not isinstance(reference, dict) or set(reference) != {REFERENCE_KEY}:
                    continue
                decoded: dict[str, Any] = {}
                for group, index in zip(GROUPS, reference[REFERENCE_KEY]):
                    decoded.update(copy.deepcopy(tables[group][index]))
                sources[component] = decoded
    return document


def _read_dmi_bulk_document(
    path: pathlib.Path | str,
    *,
    optional: bool = False,
    expand_sources: bool = True,
    allow_large_legacy: bool = False,
) -> tuple[dict[str, Any], bool]:
    source = pathlib.Path(path)
    try:
        info = source.lstat()
    except FileNotFoundError:
        if optional:
            return {}, False
        raise
    if not stat.S_ISREG(info.st_mode) or source.is_symlink() \
            or not 1 < info.st_size <= MAX_LEGACY_BYTES:
        raise ValueError("DMI bulk input is not a bounded regular file")
    if info.st_size > MAX_STORED_BYTES and not allow_large_legacy:
        raise ValueError("Large legacy DMI input requires explicit migration mode")
    with source.open("r", encoding="utf-8") as handle:
        value = json.load(handle, parse_constant=_reject_non_finite)
    if not isinstance(value, dict):
        raise ValueError("DMI bulk input must be an object")
    if "storageSchema" in value:
        if info.st_size > MAX_STORED_BYTES:
            raise ValueError("DMI encoded input exceeds its stored byte bound")
        return (
            decode_dmi_bulk_wrapper(
                value, stored_bytes=info.st_size, expand_sources=expand_sources,
            ),
            True,
        )
    if not isinstance(value.get("zones"), dict):
        raise ValueError("DMI legacy bulk input has no zones object")
    return value, False


def read_dmi_bulk_document(
    path: pathlib.Path | str,
    *,
    optional: bool = False,
    expand_sources: bool = True,
    allow_large_legacy: bool = False,
) -> dict[str, Any]:
    document, _encoded = _read_dmi_bulk_document(
        path,
        optional=optional,
        expand_sources=expand_sources,
        allow_large_legacy=allow_large_legacy,
    )
    return document


def write_dmi_bulk_document(
    path: pathlib.Path | str,
    document: dict[str, Any],
) -> int:
    destination = pathlib.Path(path)
    destination.parent.mkdir(parents=True, exist_ok=True)
    temporary = destination.with_name(destination.name + ".tmp")
    try:
        with _encoded_view(document) as (wrapper, _count):
            with temporary.open("w", encoding="utf-8", newline="\n") as handle:
                json.dump(
                    wrapper, handle, ensure_ascii=False, sort_keys=True,
                    separators=(",", ":"), allow_nan=False,
                )
                handle.write("\n")
                handle.flush()
                os.fsync(handle.fileno())
        if temporary.stat().st_size > MAX_STORED_BYTES:
            raise ValueError("DMI encoded output exceeds its stored byte bound")
        temporary.replace(destination)
        return destination.stat().st_size
    finally:
        if temporary.exists():
            temporary.unlink()


def materialize_dmi_bulk_document(
    source_path: pathlib.Path | str,
    destination_path: pathlib.Path | str,
) -> dict[str, int | bool]:
    """Validate one bounded input and atomically materialize a normal-size copy.

    A large legacy document is source-dictionary encoded.  An already encoded
    document is copied byte-for-byte after full wrapper validation.  The source
    and destination must differ so a failed migration can never damage the
    restored cache.
    """
    source = pathlib.Path(source_path)
    destination = pathlib.Path(destination_path)
    if source.resolve() == destination.resolve():
        raise ValueError("DMI storage materialization requires a separate output")
    input_bytes = source.lstat().st_size
    document, encoded_input = _read_dmi_bulk_document(
        source,
        expand_sources=False,
        allow_large_legacy=True,
    )
    if encoded_input:
        destination.parent.mkdir(parents=True, exist_ok=True)
        temporary = destination.with_name(destination.name + ".tmp")
        try:
            with source.open("rb") as reader, temporary.open("wb") as writer:
                shutil.copyfileobj(reader, writer, length=1024 * 1024)
                writer.flush()
                os.fsync(writer.fileno())
            if not 1 < temporary.stat().st_size <= MAX_STORED_BYTES:
                raise ValueError("DMI encoded copy exceeds its stored byte bound")
            temporary.replace(destination)
        finally:
            if temporary.exists():
                temporary.unlink()
    else:
        write_dmi_bulk_document(destination, document)
    read_dmi_bulk_document(destination, expand_sources=False)
    output_bytes = destination.stat().st_size
    return {
        "inputBytes": input_bytes,
        "outputBytes": output_bytes,
        "legacyNormalized": not encoded_input,
    }
